rs/http_utils/src/file_downloader.rs

```diff
@@ -182,17 +184,58 @@ pub fn check_file_hash(path: &Path, expected_sha256_hex: &str) -> FileDownloadRe
     }
 }
 
-/// Extract the contents of a given .tar.gz file into `target_dir`
-pub fn extract_tar_gz_into_dir(tar_gz_path: &Path, target_dir: &Path) -> FileDownloadResult<()> {
-    let map_to_untar_error = |e| FileDownloadError::untar_error(tar_gz_path, e);
+/// Check if the file is in Gzip format by verifying the first 2 bytes
+fn is_gz_file<R: Read>(reader: &mut R) -> io::Result<bool> {
+    let mut buffer = [0; 2];
+    reader.read_exact(&mut buffer)?;
+    Ok(buffer == [0x1f, 0x8b])
+}
+
+/// Check if the file is in Zstandard format by verifying the first 4 bytes
+fn is_zst_file<R: Read>(reader: &mut R) -> io::Result<bool> {
+    let mut buffer = [0; 4];
+    reader.read_exact(&mut buffer)?;
+    Ok(buffer == [0x28, 0xb5, 0x2f, 0xfd])
+}
 
-    let tar_gz_file =
-        File::open(tar_gz_path).map_err(|e| FileDownloadError::file_open_error(tar_gz_path, e))?;
+/// Extract the contents of a given .tar.gz or tar.zst file into `target_dir`
+pub fn extract_tar_into_dir(tar_path: &Path, target_dir: &Path) -> FileDownloadResult<()> {
+    let map_to_untar_error = |e| FileDownloadError::untar_error(tar_path, e);
+
+    let tar_file =
+        File::open(tar_path).map_err(|e| FileDownloadError::file_open_error(tar_path, e))?;
+    let mut buf_reader = BufReader::new(tar_file);
+
+    // Check for gzip format
+    buf_reader
+        .seek(SeekFrom::Start(0))
+        .map_err(map_to_untar_error)?;
+    if is_gz_file(&mut buf_reader).map_err(map_to_untar_error)? {
```

Comment:
Have you considered moving the code for seeking to the beginning into the is_XXX functions? Then you could remove all the seeking from this function which would make it slightly easier to read. In addition, it's also less surprising (to me at least) if the check doesn't advance the cursor.

Reply:
Great note! Seeks moved inside the is_XXX functions, much more readable now.